### PR TITLE
Add 3D heisenberg group in vector representation and unit tests

### PR DESCRIPTION
**Pull request by mortenapedersen**

### CODE REVIEW

File: =2.1.2 [hunk 1]
```diff
@@ -0,0 +1,75 @@
+Collecting tensorflow
```
Comment (nguigs) [inline =2.1.2:1]: This file should not be included!

Comment (nguigs) [inline =2.1.2:1]: This file should not be included

File: geomstats/geometry/heisenberg.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
+"""Implementation of the 3D Heisenberg group."""
+
+import geomstats.backend as gs
+from geomstats.geometry.euclidean import Euclidean
+from geomstats.geometry.lie_group import LieGroup
+
+
+class heisenbergVectors(LieGroup):
```
Comment (nguigs) [inline geomstats/geometry/heisenberg.py:8]: Use a capital first letter for class names

File: geomstats/geometry/heisenberg.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        """Get the identity of the 3D Heisenberg group.
+
+        Parameters
+        ----------
+        point_type : str, {'vector', 'matrix'}
+            Point_type of the returned value. Unused here.
+
+        Returns
+        -------
+        get_identity : array-like, shape=[1,]
```
Comment (nguigs) [inline geomstats/geometry/heisenberg.py:40]: `shape=[3,]` ?

File: geomstats/geometry/heisenberg.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+        return belongs
+
+    def compose(self, point_a, point_b):
+        """Compute the group product of elements `point_a` and `point_b`.
+
+        Parameters
+        ----------
+        point_a : array-like, shape=[..., {dim, [n, n]}]
+            Left factor in the product.
+        point_b : array-like, shape=[..., {dim, [n, n]}]
```
Comment (nguigs) [inline geomstats/geometry/heisenberg.py:74]: wrong shape here

File: geomstats/geometry/heisenberg.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+            Right factor in the product.
+
+        Returns
+        -------
+        composed : array-like, shape=[..., {dim, [n, n]}]
```
Comment (nguigs) [inline geomstats/geometry/heisenberg.py:79]: shape

File: geomstats/geometry/heisenberg.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+            Product of point_a and point_b along the first dimension.
+        """
+        return (point_a + point_b +
+                1 / 2 * gs.array([0, 0, point_a[0] *
+                                  point_b[1] - point_a[1] *
+                                  point_b[0]]))
```
Comment (nguigs) [inline geomstats/geometry/heisenberg.py:85]: this does not look vectorized? Consider that point_a could be a list of inputs -> use point_a[..., 0]
Can this term be computed using det or cross product?

File: geomstats/geometry/heisenberg.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+            Point.
+
+        Returns
+        -------
+        tangent_vec : array-like, shape=[..., dimension]
+            Group logarithm.
+        """
+        return point
+
+    def upperTriangular_matrix_from_vector(self, point):
```
Comment (nguigs) [inline geomstats/geometry/heisenberg.py:185]: I think this method is already implemented in the backend, or will be added by #1142

**Review by nguigs — CHANGES_REQUESTED** [review]
Great start!
- delete result.png and the file =2.1.2
- vectorize your code
- some typos in the docstrings

File: geomstats/geometry/heisenberg.py (hunk 7 above)
Comment (mortenapedersen) [reply geomstats/geometry/heisenberg.py:185]: Should I then keep this method ('upperTriangular_matrix_from_vector') in the Heisenberg class, but use the implementation from the backend or #1142, or should the method not be part of the Heisenberg class at all..?

Many thanks for you quick feedback! I'm doing the corrections now

Comment (mortenapedersen) [reply geomstats/geometry/heisenberg.py:185]: Just to be sure:   upperTriangular_matrix_from_vector   maps the vector 'point' to the matrix 

[[1, point[0], point[2] + 1 / 2 * point[0] * point[1]],
 [0, 1, point[1]],
 [0, 0, 1]] 

Are you sure that is already implemented in backend/#1142 ?

**Comment by mortenapedersen** [comment]
Hi Nicolas,

I just pushed a new version with corrected mistakes. I hope you can see the newest commit, and that I don't have to make some kind of new PR(?).

3 comments:

- the code checker complains about missing docstrings in the test script. However the other test scripts in tests_geomstats doesn't have docstrings - should I add it or not?

- I vectorized the code in compose(point_a, point_b), such that point_a can be of shape [..., 3], whereas point_b is assume to be of shape [,3] ... I hope this is correct (?).

- In compose(point_a, point_b) I found that line 91 coerces the result from float to integer, if point_ab is an integer even tough the RHS isn't. My solution to this is to convert point_a and point_b to float64's (lines 82, 83) - maybe this is a bad idea?

See you tomorrow!

File: geomstats/geometry/heisenberg.py (hunk 7 above)
Comment (nguigs) [reply geomstats/geometry/heisenberg.py]: Ah ok, no there is the method `SymmetricMatrices.from_vector` that maps `[1, a, b, 1, c, 1]` to matrix `[[1, a, b], [a, 1, c], [b, c, 1]]` and that is vectorized to work with multiple outputs, and then there is `triu` to keep only the upper part.

Comment (mortenapedersen) [reply geomstats/geometry/heisenberg.py]: Ok thanks I'll look at that, and vectorize it

**Comment by nguigs** [comment]
> I just pushed a new version with corrected mistakes. I hope you can see the newest commit, and that I don't have to make some kind of new PR(?).

No they appear here!

> 
> 3 comments:
> 
>     * the code checker complains about missing docstrings in the test script. However the other test scripts in tests_geomstats doesn't have docstrings - should I add it or not?

No don't worry, these won't show if there are no other relevant errors.

>     * I vectorized the code in compose(point_a, point_b), such that point_a can be of shape [..., 3], whereas point_b is assume to be of shape [,3] (a single point) ... I hope this is correct (?).

All cases should be considered, multiple point_a, multiple point_b, single point_a multiple point_b...

>     * In compose(point_a, point_b) I found that line 91 coerces the result from float to integer, if point_ab is an integer even tough the RHS isn't. My solution to this is to convert point_a and point_b to float64's (lines 82, 83) - maybe this is a bad idea?

Not sure to understand this one, do you mean that point_a * point_b is coerced to integer although point_a and point_b are not?

**Comment by mortenapedersen** [comment]
The latest commit includes all the corrections we have talked about so far!

File: .pre-commit-config.yaml [hunk 8]
```diff
@@ -2,7 +2,7 @@ default_language_version :
   python : python3
 repos:
   - repo: https://github.com/pre-commit/pre-commit-hooks
-    rev: v3.1.0
+    rev: v4.0.1
```
Comment (ninamiolane) [inline .pre-commit-config.yaml:5]: This file should not be in this PR.

File: geomstats/geometry/heisenberg.py [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+        """Get the identity of the 3D Heisenberg group.
+
+        Parameters
+        ----------
+        point_type : str, {'vector', 'matrix'}
+            Point_type of the returned value. Unused here.
+
+        Returns
+        -------
+        get_identity : array-like, shape=[3,]
```
Comment (ninamiolane) [inline geomstats/geometry/heisenberg.py:40]: NIT:
_ : array-like, shape=[3,]

File: geomstats/geometry/heisenberg.py [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+        Parameters
+        ----------
+        point_a : array-like, shape=[..., 3]
+            Left factor in the product.
+        point_b : array-like, shape=[..., 3]
+            Right factor in the product.
+
+        Returns
+        -------
+        composed : array-like, shape=[..., 3]
```
Comment (ninamiolane) [inline geomstats/geometry/heisenberg.py:79]: composed -> point_ab

File: geomstats/geometry/heisenberg.py [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+        """Compute the group inverse of point.
+
+        Parameters
+        ----------
+        point : array-like, shape=[..., 3]
+            Point.
+
+        Returns
+        -------
+        inv_point : array-like, shape=[..., 3]
```
Comment (ninamiolane) [inline geomstats/geometry/heisenberg.py:103]: _ : array-like, shape=

File: geomstats/geometry/heisenberg.py [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+            Indicate whether to calculate the differential of left or right
+            translations.
+            Optional, default: 'left'.
+
+        Returns
+        -------
+        jacobian : array-like, shape=[..., 3, 3]
+            Jacobian of the left/right translation by point.
+        """
+        E31 = gs.array_from_sparse([(2, 0)], [1.], (3, 3))
```
Comment (ninamiolane) [inline geomstats/geometry/heisenberg.py:130]: No uppercase letters in variable names, see:
https://www.python.org/dev/peps/pep-0008/#function-and-variable-names

File: geomstats/geometry/heisenberg.py [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+        point : array-like, shape=[..., 3]
+            Point.
+        left_or_right : str, {'left', 'right'}
+            Indicate whether to calculate the differential of left or right
+            translations.
+            Optional, default: 'left'.
+
+        Returns
+        -------
+        jacobian : array-like, shape=[..., 3, 3]
```
Comment (ninamiolane) [inline geomstats/geometry/heisenberg.py:127]: _ : array-like

File: geomstats/geometry/heisenberg.py [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+        """Compute the group exponential of the tangent vector at the identity.
+
+        Parameters
+        ----------
+        tangent_vec : array-like, shape=[..., 3]
+            Tangent vector at base point.
+
+        Returns
+        -------
+        point : array-like, shape=[..., 3]
```
Comment (ninamiolane) [inline geomstats/geometry/heisenberg.py:174]: _ :

File: geomstats/geometry/heisenberg.py [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+        """Compute the group logarithm of the point at the identity.
+
+        Parameters
+        ----------
+        point : array-like, shape=[..., 3]
+            Point.
+
+        Returns
+        -------
+        tangent_vec : array-like, shape=[..., 3]
```
Comment (ninamiolane) [inline geomstats/geometry/heisenberg.py:189]: _ :

File: tests/tests_geomstats/test_heisenberg.py [hunk 16]
```diff
@@ -0,0 +1,47 @@
+"""Unit tests for the 3D heisenberg group in vector representation."""
+
+import geomstats.backend as gs
+import geomstats.tests
+from geomstats.geometry.heisenberg import heisenbergVectors
```
Comment (ninamiolane) [inline tests/tests_geomstats/test_heisenberg.py:5]: HeisenbergVectors

File: tests/tests_geomstats/test_heisenberg.py [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+
+
+class TestHeisenbergVectors(geomstats.tests.TestCase):
+    def setUp(self):
+        self.dimension = 3
+        self.group = heisenbergVectors()
```
Comment (ninamiolane) [inline tests/tests_geomstats/test_heisenberg.py:11]: HeisenbergVectors

**Review by ninamiolane — CHANGES_REQUESTED** [review]
Excellent, just a few cosmetic comments 🌟 thanks!

File: geomstats/geometry/heisenberg.py [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+    No parameters
+
+    Reference
+    ---------
+    https://en.wikipedia.org/wiki/Heisenberg_group
+    """
+
+    def __init__(self, **kwargs):
+        super(HeisenbergVectors, self).__init__(
+            dim=int(3), default_point_type='vector',
```
Comment (ninamiolane) [inline geomstats/geometry/heisenberg.py:27]: int(3) can be just written as 3

**Review by ninamiolane — APPROVED** [review]
Great, thank you! I'm just wondering why thes tests do not run...

File: tests/tests_geomstats/test_heisenberg.py [hunk 19]
```diff
@@ -0,0 +1,55 @@
+"""Unit tests for the 3D heisenberg group in vector representation."""
+
+import geomstats.backend as gs
+import geomstats.tests
+from geomstats.geometry.heisenberg import HeisenbergVectors
+
+
+class TestHeisenbergVectors(geomstats.tests.TestCase):
+    def setUp(self):
```
Comment (ninamiolane) [inline tests/tests_geomstats/test_heisenberg.py:9]: this should be setup_method, see: https://github.com/geomstats/geomstats/blob/6c3ed380c992d7f866448845e890ed198ce8bbea/tests/tests_geomstats/test_connection.py#L16

**Review by ninamiolane — CHANGES_REQUESTED** [review]

**Pull request merged**